File: frontend/src/services/messageService.ts
import api from './api';
import type {
  ApiResponse,
  CursorPageResponse,
  MarkAsReadRequest,
  Message,
  SendMessageRequest,
  UpdateMessageRequest,
} from '@/types';

export const messageService = {
  async getMessages(
    conversationId: string,
    cursor?: string,
    limit = 30
  ): Promise<CursorPageResponse<Message>> {
    const response = await api.get<ApiResponse<CursorPageResponse<Message>>>(
      `/conversations/${conversationId}/messages`,
      {
        params: { cursor, limit },
      }
    );
    return response.data.data;
  },

  async sendMessage(
    conversationId: string,
    data: SendMessageRequest
  ): Promise<Message> {
    const response = await api.post<ApiResponse<Message>>(
      `/conversations/${conversationId}/messages`,
      data
    );
    return response.data.data;
  },

  async updateMessage(
    messageId: string,
    data: UpdateMessageRequest
  ): Promise<Message> {
    const response = await api.put<ApiResponse<Message>>(
      `/messages/${messageId}`,
      data
    );
    return response.data.data;
  },

  async deleteMessage(messageId: string): Promise<void> {
    await api.delete<ApiResponse<void>>(`/messages/${messageId}`);
  },

  async markAsRead(
    conversationId: string,
    data?: MarkAsReadRequest
  ): Promise<void> {
    await api.post<ApiResponse<void>>(
      `/conversations/${conversationId}/read`,
      data
    );
  },
};

File: backend/src/main/java/com/chatweb/message/service/MessageService.java
package com.chatweb.message.service;

import com.chatweb.common.dto.CursorPageResponse;
import com.chatweb.message.dto.MarkAsReadRequest;
import com.chatweb.message.dto.MessageResponse;
import com.chatweb.message.dto.MessageSummaryResponse;
import com.chatweb.message.dto.SendMessageRequest;
import com.chatweb.message.dto.UpdateMessageRequest;

import java.util.Collection;
import java.util.Map;
import java.util.UUID;

public interface MessageService {

    MessageResponse sendMessage(UUID currentUserId, UUID conversationId, SendMessageRequest request);

    CursorPageResponse<MessageResponse> getMessages(UUID currentUserId, UUID conversationId, String cursor, int limit);

    MessageResponse updateMessage(UUID currentUserId, UUID messageId, UpdateMessageRequest request);

    void deleteMessage(UUID currentUserId, UUID messageId);

    void markAsRead(UUID currentUserId, UUID conversationId, MarkAsReadRequest request);

    MessageSummaryResponse getLatestMessage(UUID conversationId);

    Map<UUID, MessageSummaryResponse> getLatestMessages(Collection<UUID> conversationIds);

    long getUnreadCount(UUID conversationId, UUID userId, UUID lastReadMessageId);
}
